### ISSUE TITLE
Support `@psalm-type` and `@psalm-import-type` for importing complex types from other symbols?

### ISSUE THREAD

**Issue by Ocramius** [CuyZ/Valinor#26]
## Background

Usage of `@psalm-type` and `@psalm-import-type` allows declaring local and foreign aliases for complex type compositions.

These types ares extremely useful when creating a large trees of array-like structures, where creating a large tree of objects would potentially be worse/wasteful from a development perspective.

Most often, these are useful when attempting to map object structures to array structures that vary slightly, or for supporting array keys that would otherwise be illegal in PHP (such as spaces in array keys, or special languages).

## Related issues

This is a follow-up of:
 * #18 
 * #20 

## Example / Test Case 

Here's a stripped down example with a crash, useful for evaluating whether the feature is "working", if/when we work on it:

```php
<?php

namespace My\App;

use CuyZ\Valinor\MapperBuilder;

require_once __DIR__ . '/Valinor/vendor/autoload.php';

/**
 * @psalm-type ImportedType = array{
 *   importedType: non-empty-string
 * }
 */
class SomeClassThatDeclaresAliases
{
}

/**
 * @psalm-import-type ImportedType from SomeClassThatDeclaresAliases
 *
 * @psalm-type LocalType = array{
 *   localType: int
 * }
 */
class MyArrayWrapper
{
    /** @var array{
     *     0: LocalType
     *     1: ImportedType
     * }
     */
    public array $value;
}

$mapper = (new MapperBuilder())
    ->withCacheDir(__DIR__ . '/cache')
    ->mapper();

var_dump($mapper->map(
    MyArrayWrapper::class,
    ['value' => [
        0 => ['localType' => 123],
        1 => ['importedType' => 'here be dragons']
    ]]
));
```

Currently crashes due to unknown types:

```
php example-18-imported-types.php 

Fatal error: Uncaught CuyZ\Valinor\Type\Types\UnresolvableType: The type `array{     0: LocalType     1: ImportedType }` for property `My\App\MyArrayWrapper::$value` could not be resolved: Cannot parse unknown symbol `LocalType`. in /app/Valinor/src/Definition/Repository/Reflection/ReflectionPropertyDefinitionBuilder.php:137
Stack trace:
#0 /app/Valinor/src/Definition/Repository/Reflection/ReflectionPropertyDefinitionBuilder.php(110): CuyZ\Valinor\Definition\Repository\Reflection\ReflectionPropertyDefinitionBuilder->parseType('array{     0: L...', Object(ReflectionProperty), Object(CuyZ\Valinor\Type\Parser\LexingParser))
#1 /app/Valinor/src/Definition/Repository/Reflection/ReflectionPropertyDefinitionBuilder.php(70): CuyZ\Valinor\Definition\Repository\Reflection\ReflectionPropertyDefinitionBuilder->typeFromDocBlock(Object(CuyZ\Valinor\Definition\ClassSignature), Object(ReflectionProperty))
#2 /app/Valinor/src/Definition/Repository/Reflection/ReflectionPropertyDefinitionBuilder.php(43): CuyZ\Valinor\Definition\Repository\Reflection\ReflectionPropertyDefinitionBuilder->resolveType(Object(CuyZ\Valinor\Definition\ClassSignature), Object(ReflectionProperty))
#3 /app/Valinor/src/Definition/Repository/Reflection/ReflectionClassDefinitionRepository.php(41): CuyZ\Valinor\Definition\Repository\Reflection\ReflectionPropertyDefinitionBuilder->for(Object(CuyZ\Valinor\Definition\ClassSignature), Object(ReflectionProperty))
#4 [internal function]: CuyZ\Valinor\Definition\Repository\Reflection\ReflectionClassDefinitionRepository->CuyZ\Valinor\Definition\Repository\Reflection\{closure}(Object(ReflectionProperty))
#5 /app/Valinor/src/Definition/Repository/Reflection/ReflectionClassDefinitionRepository.php(42): array_map(Object(Closure), Array)
#6 /app/Valinor/src/Definition/Repository/Cache/CacheClassDefinitionRepository.php(36): CuyZ\Valinor\Definition\Repository\Reflection\ReflectionClassDefinitionRepository->for(Object(CuyZ\Valinor\Definition\ClassSignature))
#7 /app/Valinor/src/Mapper/Tree/Builder/ClassNodeBuilder.php(35): CuyZ\Valinor\Definition\Repository\Cache\CacheClassDefinitionRepository->for(Object(CuyZ\Valinor\Definition\ClassSignature))
#8 /app/Valinor/src/Mapper/Tree/Builder/CasterNodeBuilder.php(35): CuyZ\Valinor\Mapper\Tree\Builder\ClassNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#9 /app/Valinor/src/Mapper/Tree/Builder/VisitorNodeBuilder.php(28): CuyZ\Valinor\Mapper\Tree\Builder\CasterNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#10 /app/Valinor/src/Mapper/Tree/Builder/ValueAlteringNodeBuilder.php(31): CuyZ\Valinor\Mapper\Tree\Builder\VisitorNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#11 /app/Valinor/src/Mapper/Tree/Builder/ShellVisitorNodeBuilder.php(30): CuyZ\Valinor\Mapper\Tree\Builder\ValueAlteringNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#12 /app/Valinor/src/Mapper/Tree/Builder/ErrorCatcherNodeBuilder.php(23): CuyZ\Valinor\Mapper\Tree\Builder\ShellVisitorNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#13 /app/Valinor/src/Mapper/Tree/Builder/RootNodeBuilder.php(21): CuyZ\Valinor\Mapper\Tree\Builder\ErrorCatcherNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#14 /app/Valinor/src/Mapper/TreeMapperContainer.php(57): CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell))
#15 /app/Valinor/src/Mapper/TreeMapperContainer.php(31): CuyZ\Valinor\Mapper\TreeMapperContainer->node('My\\App\\MyArrayW...', Array)
#16 /app/example-18-imported-types.php(42): CuyZ\Valinor\Mapper\TreeMapperContainer->map('My\\App\\MyArrayW...', Array)
#17 {main}

Next CuyZ\Valinor\Mapper\Tree\Exception\UnresolvableShellType: The type `array{     0: LocalType     1: ImportedType }` for property `My\App\MyArrayWrapper::$value` could not be resolved: Cannot parse unknown symbol `LocalType`. in /app/Valinor/src/Mapper/Tree/Shell.php:40
Stack trace:
#0 /app/Valinor/src/Mapper/Tree/Shell.php(57): CuyZ\Valinor\Mapper\Tree\Shell->__construct(Object(CuyZ\Valinor\Type\Types\UnresolvableType), Array)
#1 /app/Valinor/src/Mapper/Tree/Builder/ClassNodeBuilder.php(41): CuyZ\Valinor\Mapper\Tree\Shell->child('value', Object(CuyZ\Valinor\Type\Types\UnresolvableType), Array, Object(CuyZ\Valinor\Definition\NativeAttributes))
#2 /app/Valinor/src/Mapper/Tree/Builder/CasterNodeBuilder.php(35): CuyZ\Valinor\Mapper\Tree\Builder\ClassNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#3 /app/Valinor/src/Mapper/Tree/Builder/VisitorNodeBuilder.php(28): CuyZ\Valinor\Mapper\Tree\Builder\CasterNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#4 /app/Valinor/src/Mapper/Tree/Builder/ValueAlteringNodeBuilder.php(31): CuyZ\Valinor\Mapper\Tree\Builder\VisitorNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#5 /app/Valinor/src/Mapper/Tree/Builder/ShellVisitorNodeBuilder.php(30): CuyZ\Valinor\Mapper\Tree\Builder\ValueAlteringNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#6 /app/Valinor/src/Mapper/Tree/Builder/ErrorCatcherNodeBuilder.php(23): CuyZ\Valinor\Mapper\Tree\Builder\ShellVisitorNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#7 /app/Valinor/src/Mapper/Tree/Builder/RootNodeBuilder.php(21): CuyZ\Valinor\Mapper\Tree\Builder\ErrorCatcherNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell), Object(CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder))
#8 /app/Valinor/src/Mapper/TreeMapperContainer.php(57): CuyZ\Valinor\Mapper\Tree\Builder\RootNodeBuilder->build(Object(CuyZ\Valinor\Mapper\Tree\Shell))
#9 /app/Valinor/src/Mapper/TreeMapperContainer.php(31): CuyZ\Valinor\Mapper\TreeMapperContainer->node('My\\App\\MyArrayW...', Array)
#10 /app/example-18-imported-types.php(42): CuyZ\Valinor\Mapper\TreeMapperContainer->map('My\\App\\MyArrayW...', Array)
#11 {main}
  thrown in /app/Valinor/src/Mapper/Tree/Shell.php on line 40

```

**Comment by romm** [comment]
Hi @Ocramius, I did a first run to handle the local type aliasing feature in #29. Can you give it a try?

Next will be the type alias importing. 🙁

**Comment by romm** [comment]
Aaand there you go for the type alias imports. 🙂 

See #29 as well!

**Comment by romm** [comment]
Hi @Ocramius, I just released v0.2.0 shipped with these features. If/when you have time to check it out, please report any problem found in a new GH issue. 🙂
